### ISSUE TITLE
# GET_CERTIFICATE/CERTIFICATE issue (from mjunior <mjunior@larc.usp.br>)

### ISSUE THREAD

**Issue by jyao1** [jyao1/openspdm#91]
From mjunior <mjunior@larc.usp.br>

## Potential OpenSPDM code issues:
  - Message transcript B buffer may overflow on long certificate chains or if the certificate chain is retrieved in small chunks. Currently, the overflow is caught by an ASSERT statement, while the return value of buffer appendage functions is not checked. Fix suggestion: check the return value of buffer appendage and drop the ASSERT statements. Nonetheless, as mentioned previously, the standard does not define what to do in case of buffer overflow.
## Contribution to OpenSPDM:
New tests added.
  - On TestSpdmRequesterGetCertificate.c:
        [PASS] Invalid certificate signature
        [PASS] Failed certificate chain check
        [PASS] Get a certificate chain that fits in one single message
        [FAIL] Get certificate chain byte by byte => message transcript B buffer overflows, fires ASSERT
        [FAIL] Get a long certificate chain => message transcript B buffer overflows, fires ASSERT
  - On TestSpdmResponderCertificate.c:
        [PASS] Tests varying length, while offset = 0
        [PASS] Tests varying offset, while length = 0
        [PASS] Tests varying length and offset
        [PASS] Tests large certificate chains boundary values of Offset and Length
        [PASS] Certificate fits in one single message
        [FAIL] Requests byte by byte => message transcript B buffer overflows, fires ASSERT

**Comment by jyao1** [comment]
# TestSpdmResponderCertificateCase11() fails with MbedTls.

## [Root cause]
This test case will read Shorter1024B_bundle_responder.certchain.der.
But it fails at SpdmVerifyCertificateBuffer(), because mbedTls x509_crt_parse_der_core() need check the issuer name.
```
    if( ( ret = mbedtls_x509_get_name( &p, p + len, &crt->issuer ) ) != 0 )
    {
        mbedtls_x509_crt_free( crt );
        return( ret );
    }

    crt->issuer_raw.len = p - crt->issuer_raw.p;
```

The cert dump is below. It also used an unsupported RSA768 algorithm.
```
jyao1@jyao1-MOBL:~/openspdm/openspdm/OsTest/TestKey/LongChains$ openssl x509 -in Shorter1024B_bundle_responder.certchain.der -inform der -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            22:9b:92:51:3c:53:7b:9a:5c:7b:24:1d:5c:21:f8:ce:8e:57:ca:87
        Signature Algorithm: sha256WithRSAEncryption
        Issuer:
        Validity
            Not Before: Nov 18 18:48:08 2020 GMT
            Not After : Nov 16 18:48:08 2030 GMT
        Subject:
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (768 bit)
                Modulus:
                    00:d2:ed:9c:66:2d:15:c4:ff:d7:a8:12:32:17:2d:
                    6e:a0:b9:91:8a:aa:ef:17:c2:17:70:e1:58:cb:8a:
                    94:7d:3e:25:ce:a2:65:1e:0f:5b:7e:b0:69:63:48:
                    24:25:3a:06:4e:71:86:7f:de:d2:4c:6b:5a:b4:fc:
                    45:4e:f5:91:ba:65:19:da:03:60:a5:8a:cd:18:c2:
                    b8:e8:39:bd:bf:58:62:a6:0b:ad:24:01:1b:63:7d:
                    eb:f4:0f:3d:43:3b:b1
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Subject Key Identifier:
                D8:D1:6E:C5:7E:C7:F0:2A:7C:88:97:78:BD:06:70:2F:81:EF:07:1F
            X509v3 Authority Key Identifier:
                keyid:D8:D1:6E:C5:7E:C7:F0:2A:7C:88:97:78:BD:06:70:2F:81:EF:07:1F

            X509v3 Basic Constraints: critical
                CA:TRUE
    Signature Algorithm: sha256WithRSAEncryption
         21:5b:f0:7a:b0:15:18:d1:c7:39:79:62:d5:d8:82:19:aa:de:
         8b:4b:e1:b0:39:83:31:8e:3b:b6:c1:ef:d0:ee:c8:1d:c6:08:
         f0:ed:c7:42:e4:10:70:6f:f7:fd:aa:6e:11:11:f0:68:d8:37:
         9b:2e:7a:ef:8a:0b:91:da:2f:5b:dc:01:45:a5:12:af:b7:06:
         8b:c5:1b:75:64:df:c8:f4:1c:3f:c4:28:97:90:11:a1:25:51:
         10:ab:71:90:c1:6c
```

## [Recommendation]
Test case fix: (https://github.com/jyao1/openspdm/issues/133)
Please use RSA2048 at least. 
Please give issuer name.
Please test with MbedTls.

**Comment by jyao1** [comment]
# TestSpdmRequesterGetCertificateCase13() fails with MbedTls

## [Root cause]
This test case will read Shorter1024B_bundle_requester.certchain.der.
It does not have issuer name.
Same root-cause as above.

## [Recommendation]
Test case fix: (https://github.com/jyao1/openspdm/issues/133)
Please use RSA2048 at least.
Please give issuer name.
Please test with MbedTls.

**Comment by jyao1** [comment]
# TestSpdmResponderCertificateCase12() assert in AppendManagedBuffer()

## [Root Cause]
This test case requests 1 byte one time.
It causes GetCertificate/Certificate message 1348 times.

All those messages are be appended in MessageB buffer, which is 4096 as maximum.
It assert in AppendManagedBuffer() 
```
  if (BufferSize > ManagedBuffer->MaxBufferSize - ManagedBuffer->BufferSize) {
    ASSERT (FALSE);
    return RETURN_BUFFER_TOO_SMALL;
  }
```

## [Recommendation]
openspdm fix: (https://github.com/jyao1/openspdm/issues/132)
Need remove ASSERT (FALSE); from AppendManagedBuffer() on RETURN_BUFFER_TOO_SMALL.
Need handle AppendManagedBuffer failure in each commend parser. If there are too many request, openspdm may return error.

Test case fix:
Should not always assume command success.
Too many small cert request may be considered as an attack.

**Comment by jyao1** [comment]
# TestSpdmRequesterGetCertificateCase14() assert in AppendManagedBuffer()

## [Root case]
This test case requests 1 byte one time.
Same root case as above.

## [Recommendation]
Same as above.
openspdm fix: (#132)
Need remove ASSERT (FALSE); from AppendManagedBuffer() on RETURN_BUFFER_TOO_SMALL.
Need handle AppendManagedBuffer failure in each commend parser. If there are too many request, openspdm may return error.

Test case fix:
Should not always assume command success.
Too many small cert request may be considered as an attack.

**Comment by jyao1** [comment]
# TestSpdmRequesterGetCertificateCase15() assert in AppendManagedBuffer()

## [Root Cause]
This test case uses a long certificate 65210.

The messages are be appended in MessageB buffer, which is 4096 as maximum.
It assert in AppendManagedBuffer()
```
  if (BufferSize > ManagedBuffer->MaxBufferSize - ManagedBuffer->BufferSize) {
    ASSERT (FALSE);
    return RETURN_BUFFER_TOO_SMALL;
  }
```

## [Recommendation]
openspdm fix: (#132)
Need remove ASSERT (FALSE); from AppendManagedBuffer() on RETURN_BUFFER_TOO_SMALL.
Need handle AppendManagedBuffer failure in each commend parser. If there are too many request, openspdm may return error.

Test case fix:
Should not always assume command success.
Unreasonable long certificate may be considered as an attack.
